src/server/middleware/blob/put_blob_upload.go

```diff
@@ -15,17 +15,53 @@
 package blob
 
 import (
-	"net/http"
-	"strconv"
-
+	"fmt"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/lib/errors"
 	"github.com/goharbor/harbor/src/lib/log"
+	blob_models "github.com/goharbor/harbor/src/pkg/blob/models"
 	"github.com/goharbor/harbor/src/pkg/distribution"
 	"github.com/goharbor/harbor/src/server/middleware"
+	"github.com/goharbor/harbor/src/server/middleware/requestid"
+	"net/http"
+	"strconv"
 )
 
-// PutBlobUploadMiddleware middleware to create Blob and ProjectBlob after PUT /v2/<name>/blobs/uploads/<session_id> success
+// PutBlobUploadMiddleware middleware to create Blob and ProjectBlob after PUT /v2/<name>/blobs/uploads/<session_id>?digest=<digest> success
 func PutBlobUploadMiddleware() func(http.Handler) http.Handler {
-	return middleware.AfterResponse(func(w http.ResponseWriter, r *http.Request, statusCode int) error {
+
+	before := middleware.BeforeRequest(func(r *http.Request) error {
+		ctx := r.Context()
+		logger := log.G(ctx)
+
+		v := r.URL.Query()
+		digest := v.Get("digest")
+
+		// digest empty is handled by the blob controller GET method
+		bb, err := blob.Ctl.Get(r.Context(), digest)
+		if err != nil {
+			if errors.IsNotFoundErr(err) {
+				return nil
+			}
+			return err
```

Comment:
Will such error yield a 4xx? or 5xx? if the request does not contain digest in query string the server should return a 4xx not 5xx

Reply:
if you go into the code of blob controller get method, if the digest is empty, it will return a 400.

Reply:
func (c *controller) Get(ctx context.Context, digest string, options ...Option) (*blob.Blob, error) {
	if digest == "" {
		return nil, errors.New(nil).WithCode(errors.BadRequestCode).WithMessage("require digest")
	}

	opts := newOptions(options...)

	params := blob.ListParams{
		ArtifactDigest: opts.ArtifactDigest,
		BlobDigests:    []string{digest},
		ProjectID:      opts.ProjectID,
	}

	blobs, err := c.blobMgr.List(ctx, params)
	if err != nil {
		return nil, err
	} else if len(blobs) == 0 {
		return nil, errors.NotFoundError(nil).WithMessage("blob %s not found", digest)
	}

	return blobs[0], nil
}

Reply:
ok